utils/include/edm4hep/utils/vector_utils.h

```diff
@@ -0,0 +1,270 @@
+#ifndef EDM4HEP_UTILS_VECTOR_HH
+#define EDM4HEP_UTILS_VECTOR_HH
+
+// These ultilies require concepts. If not available, use the fallback
+// vector_utils_legacy.h instead to capture most functionality.
+#if !__cpp_concepts
+#include <edm4hep/utils/vector_utils_legacy.h>
+#else
+
+#include <edm4hep/Vector3f.h>
+
+#include <cmath>
+
+namespace edm4hep {
+
+template <class V>
+concept VectorHasX = requires(V v) {
+  v.x;
+};
+
+template <class V>
+concept VectorHasY = requires(V v) {
+  v.y;
+};
+
+template <class V>
+concept VectorHasZ = requires(V v) {
+  v.z;
+};
+
+template <class V>
+concept VectorHasA = requires(V v) {
+  v.a;
+};
+
+template <class V>
+concept VectorHasB = requires(V v) {
+  v.b;
+};
+
+template <class V>
+concept ClassVector = requires(V v) {
+  v.x();
+};
+
+template <class V>
+concept Vector2D_XY = VectorHasX<V>&& VectorHasY<V> && !VectorHasZ<V> && !ClassVector<V>;
+
+template <class V>
+concept Vector2D_AB = VectorHasA<V>&& VectorHasB<V> && !VectorHasZ<V> && !ClassVector<V>;
+
+template <class V>
+concept Vector2D = Vector2D_XY<V> || Vector2D_AB<V>;
+
+template <class V>
+concept Vector3D = VectorHasX<V>&& VectorHasY<V>&& VectorHasZ<V> && !ClassVector<V>;
+
+template <class V>
+concept VectorND = Vector2D<V> || Vector3D<V>;
+
+template <class V>
+concept VectorND_XYZ = Vector2D_XY<V> || Vector3D<V>;
+
+namespace utils {
+  inline double etaToAngle(const double eta) {
+    return std::atan(std::exp(-eta)) * 2.;
+  }
+
+  inline double angleToEta(const double theta) {
+    return -std::log(std::tan(0.5 * theta));
+  }
+
+  // Utility getters to accomodate different vector types
+  template <VectorND_XYZ V>
+  constexpr auto vector_x(const V& v) {
+    return v.x;
+  }
+
+  template <VectorND_XYZ V>
+  constexpr auto vector_y(const V& v) {
+    return v.y;
+  }
+
+  template <Vector3D V>
+  constexpr auto vector_z(const V& v) {
+    return v.z;
+  }
+
+  template <Vector2D_AB V>
+  constexpr auto vector_x(const V& v) {
+    return v.a;
+  }
+
+  template <Vector2D_AB V>
+  constexpr auto vector_y(const V& v) {
+    return v.b;
+  }
+
+  template <Vector2D V>
+  constexpr auto vector_z(const V&) {
+    return 0;
+  }
+
+  namespace detail {
+    /// Helper struct to determine the underlying type of edm4hep vector types
+    template <typename V>
+    struct ValueTypeHelper {
+      using type = decltype(vector_x(std::declval<V>()));
+    };
+  } // namespace detail
+
+  /// Type alias that returns the underlying type of edm4hep
+  template <typename V>
+  using ValueType = typename detail::ValueTypeHelper<V>::type;
+
+  // inline edm4hep::Vector2f VectorFromPolar(const double r, const double theta)
+  // {
+  //  return {r * sin(theta), r * cos(theta)};
+  //}
+
+  template <Vector3D V = edm4hep::Vector3f>
+  V sphericalToVector(const double r, const double theta, const double phi) {
+    using FloatType = decltype(vector_x(V()));
```

Comment:
```suggestion
    using FloatType = ValueType<V>;
```
Since this is now a thing.